### DISCUSSION TITLE
Lock improvements in Swig Director for Golang

### DISCUSSION THREAD

**Discussion by bardiz12** [swig/swig#3001] [Ideas]
Hi!. I've been using swig for golang for sometimes and there's something about the Swig director features that I think we can improve.

```go
// file: /Lib/go/goruntime.swg

var swigDirectorTrack struct {
	sync.Mutex
	m map[int]interface{}
	c int
}

func swigDirectorLookup(c int) interface{} {
	swigDirectorTrack.Lock()
	defer swigDirectorTrack.Unlock()
	ret := swigDirectorTrack.m[c]
	if ret == nil {
		panic("C++ director pointer not found (possible	use-after-free)")
	}
	return ret
}
```

I notice that the swigDirectorTrack variable is using `sync.Mutex` rather than `sync.RWMutex`. I think it would be beneficial in term of performance if we can implement the swigDirectorTrack using sync.RWMutex since sync.RWMutex will allow concurrent read. If we use the `RLock()` in `swigDirectorLookup` function, it will allow concurrent read to `swigDirectorTrack.m` map.

Thank You

**Comment by ojwb** [comment]
In #3005, ianlancetaylor said about this suggestion:
> A `sync.RWMutex` is more complex than a `sync.Mutex`.  It's worth using a `sync.RWMutex` when the read lock can be held for more than a brief period of time.  But in this case the read lock is held only very briefly.  For a case like this the more efficient `sync.Mutex` is likely to outweigh the benefit of permitting multiple concurrent readers.  So I would not be inclined to make this change, unless of course there are some benchmarks showing that it would make the code run faster.  Thanks.

@bardiz12 Do you have benchmarking results to show your proposed change actually makes the code run faster?
